analyzers/ruby/ruby.go

```diff
@@ -127,47 +127,85 @@ func (a *Analyzer) IsBuilt() (bool, error) {
 }
 
 func (a *Analyzer) Analyze() (graph.Deps, error) {
-	lockfilePath := filepath.Join(a.Module.Dir, "Gemfile.lock")
-	if a.Options.LockfilePath != "" {
-		lockfilePath = a.Options.LockfilePath
+	strategy := "list"
+
+	if a.Options.Strategy != "" {
+		strategy = a.Options.Strategy
 	}
 
-	switch a.Options.Strategy {
+	return a.analyzeHelper(strategy)
+}
+
+func (a *Analyzer) analyzeHelper(strategy string) (graph.Deps, error) {
+	lockfilePath := a.getLockfilePath()
+
+	switch strategy {
 	case "list":
-		gems, err := a.Bundler.List()
-		if err != nil {
-			return graph.Deps{}, err
-		}
-		imports, deps := FromGems(gems)
-		return graph.Deps{
-			Direct:     imports,
-			Transitive: deps,
-		}, nil
+		return a.bundlerListAnalyzerStrategy()
 	case "lockfile":
-		lockfile, err := bundler.FromLockfile(lockfilePath)
-		if err != nil {
-			return graph.Deps{}, err
-		}
-		imports, deps := FromLockfile(lockfile)
-		return graph.Deps{
-			Direct:     imports,
-			Transitive: deps,
-		}, nil
+		return a.lockfileAnalyzerStrategy(lockfilePath)
 	case "list-lockfile":
 		fallthrough
 	default:
-		lockfile, err := bundler.FromLockfile(lockfilePath)
-		if err != nil {
-			return graph.Deps{}, err
-		}
-		gems, err := a.Bundler.List()
-		if err != nil {
-			return graph.Deps{}, err
-		}
-		imports, deps := FilteredLockfile(gems, lockfile)
-		return graph.Deps{
-			Direct:     imports,
-			Transitive: deps,
-		}, nil
+		return a.defaultAnalyzerStrategy(lockfilePath)
+	}
+}
+
+func (a *Analyzer) bundlerListAnalyzerStrategy() (graph.Deps, error) {
+	gems, err := a.Bundler.List()
+	if err != nil {
+		lockfilePath := a.getLockfilePath()
+
+		return a.lockfileAnalyzerStrategy(lockfilePath)
```

Comment:
I'd rather have the fallback occuring in the caller of each Analyzer strategy. This seems easier to reason about.